storage/src/nodestore.rs

```diff
@@ -755,6 +787,60 @@ impl NodeStoreHeader {
         }
     }
 
+    fn validate(&self) -> Result<(), Error> {
+        self.version.validate()?;
+        self.validate_endian_test()?;
+        self.validate_area_size_hash()?;
+        self.validate_ethhash()?;
+        Ok(())
+    }
+
+    fn validate_endian_test(&self) -> Result<(), Error> {
+        if self.endian_test == 1 {
+            Ok(())
+        } else {
+            Err(Error::new(
+                ErrorKind::InvalidData,
+                "Database cannot be opened due to difference in endianness",
+            ))
+        }
+    }
+
+    fn validate_area_size_hash(&self) -> Result<(), Error> {
+        if self.area_size_hash == area_size_hash().as_slice() {
+            Ok(())
+        } else {
+            Err(Error::new(
+                ErrorKind::InvalidData,
+                "Database cannot be opened due to difference in area size hash",
+            ))
+        }
+    }
+
+    #[cfg(not(feature = "ethhash"))]
```

Comment:
nit: almost feels like a macro or something would have been better here

Reply:
macro to check for ethhash or macro to reduce the boilerplate of the validate methods?

For checking ethhash, there's the `cfg!` macro but that would make the code a less clear. 

```rust
fn validate_ethhash(&self) -> Result<(), Error> {
    if self.ethhash == u64::from(cfg!(feature = "ethhash")) {
        Ok(())
    } else {
        Err(Error::new(
            ErrorKind::InvalidData,
            if cfg!(feature = "ethhash") {
                "Database cannot be opened as it was created without ethhash enabled"
            } else {
                "Database cannot be opened as it was created with ethhash enabled"
            },
        ))
    }
}

```

For a custom macro, there's a huge maintainability cost to benefit ratio that doesn't make it worth it.